src/main/java/org/tugraz/sysds/runtime/lineage/Lineage.java

```diff
@@ -0,0 +1,172 @@
+                    break;
+            }
+        } else if (instruction instanceof LineageTraceable) {
+            addLineageItem(((LineageTraceable) instruction).getLineageItem());
+        } else {
+            System.out.printf("Oh no! Unknown Instruction traced: %s (%s)\n", instruction.getOpcode(), instruction.getClass().getName());
+        }
+    }
+
+    public static void removeLineageItem(String key) {
+        LineageItem li = lineage_traces.get(key);
+        if (li == null)
+            return;
+
+        removeAncestorLink(li);
+        lineage_traces.remove(key);
+    }
+
+    public static void addLineageItem(LineageItem li) {
+        if (lineage_traces.get(li.getKey()) != null) {
+            removeAncestorLink(lineage_traces.get(li.getKey()));
+            lineage_traces.remove(li.getKey());
+        }
+        lineage_traces.put(li.getKey(), li);
+    }
+
+    public static LineageItem getOrCreate(CPOperand variable) {
+        if (variable == null)
+            return null;
+        if (lineage_traces.get(variable.getName()) == null)
+            return new LineageItem(variable);
+        return lineage_traces.get(variable.getName());
+    }
+
+    private static void writeInstruction(VariableCPInstruction inst, ExecutionContext ec) {
+        LineageItem li = lineage_traces.get(inst.getInput1().getName());
+        String desc = Explain.explain(li);
+
+        String fname = ec.getScalarInput(inst.getInput2().getName(), Types.ValueType.STRING, inst.getInput2().isLiteral()).getStringValue();
+        fname += ".lineage";
+
+        try {
+            HDFSTool.writeStringToHDFS(desc, fname);
+            FileSystem fs = IOUtilFunctions.getFileSystem(fname);
+            if (fs instanceof LocalFileSystem) {
+                Path path = new Path(fname);
+                IOUtilFunctions.deleteCrcFilesFromLocalFileSystem(fs, path);
+            }
+
+        } catch (IOException e) {
+            throw new DMLRuntimeException(e);
+        }
+    }
+
+    private static void removeInstruction(VariableCPInstruction inst) {
+        removeLineageItem(inst.getInput1().getName());
+    }
+
+    private static void moveInstruction(VariableCPInstruction inst) {
+        if (inst.getInput2().getName().equals("__pred")) {
+            removeLineageItem(inst.getInput1().getName());
+        } else {
+            LineageItem li_source = get(inst.getInput1());
+            LineageItem li;
+            ArrayList<LineageItem> lineages = new ArrayList<>();
+            if (li_source != null) {
+                lineages.add(li_source);
+                li = new LineageItem(inst.getInput2(), lineages, inst.getOpcode());
+            } else {
+                lineages.add(getOrCreate(inst.getInput1()));
+                if (inst.getInput3() != null)
+                    lineages.add(getOrCreate(inst.getInput3()));
+                li = new LineageItem(inst.getInput2(), lineages, inst.getOpcode());
```

Comment:
as the same LineageItem is created in both if and else branches, you can pull it out (before or after) and simply the code.